dbt_semantic_interfaces/parsing/where_filter/query_interface.py

```diff
@@ -0,0 +1,65 @@
+from __future__ import annotations
+
+from typing import Protocol, Sequence
+
+
+class QueryInterfaceDimension(Protocol):
+    """Represents the interface for Dimension in the query interface."""
+
+    def grain(self, _grain: str) -> QueryInterfaceDimension:
+        """The time granularity."""
+        raise NotImplementedError
```

Comment:
We shouldn't do `raise NotImplementedError` in the protocols. Protocols aren't actually inherited in the way that we're used to. Direct inheritance of protocols like
```
class TheImplementation(TheProtocol):
  ...
```
is actually discouraged. Weird things start happening when you do. Thus the default implementation in the protocol will not be called. We don't want to give people a false sense of security that a `NotImplementedError` will be raised, so we shouldn't make it look like it will.